analyzers/src/SonarAnalyzer.CSharp/Rules/TooManyLoggingCalls.cs

```diff
@@ -0,0 +1,217 @@
+            {
+                cc.RegisterNodeAction(Process, SyntaxKind.Block, SyntaxKind.CompilationUnit);
+            }
+        });
+
+    private void Process(SonarSyntaxNodeReportingContext context)
+    {
+        var node = context.Node;
+        if (IsBlockNodeSupported(node))
+        {
+            var logCallCollector = new LoggingCallCollector(context.SemanticModel, node);
+            logCallCollector.Visit(node);
+            foreach (var loggingCategory in logCallCollector.LoggingInvocations)
+            {
+                var treshold = Treshold(loggingCategory.Key);
+                var invocations = loggingCategory.Value;
+                if (invocations.Count > treshold)
+                {
+                    var primaryLocation = invocations[0].GetLocation();
+                    var secondaryLocations = invocations.Skip(1).Select(x => x.GetLocation());
+                    context.ReportIssue(Diagnostic.Create(Rule, primaryLocation, secondaryLocations, loggingCategory.Key, invocations.Count, treshold));
+                }
+            }
+        }
+    }
+
+    private int Treshold(string category)
+    {
+        var threshold = category switch
+        {
+            CategoryNames.Debug => DebugThreshold,
+            CategoryNames.Error => ErrorThreshold,
+            CategoryNames.Warning => WarningThreshold,
+            _ => InformationThreshold
+        };
+        return Math.Max(threshold, 0);
+    }
+
+    private static bool IsBlockNodeSupported(SyntaxNode node) =>
+        node.Kind() == SyntaxKind.Block
+        || (node.Kind() == SyntaxKind.CompilationUnit && node.ChildNodes().OfType<GlobalStatementSyntax>().Any());  // for top-level statements
+
+    private sealed class LoggingCallCollector : SafeCSharpSyntaxWalker
+    {
+        private readonly SemanticModel semanticModel;
+        private SyntaxNode currentBlock;
+
+        public Dictionary<string, List<InvocationExpressionSyntax>> LoggingInvocations { get; } = [];
+
+        public LoggingCallCollector(SemanticModel semanticModel, SyntaxNode currentBlock)
+        {
+            this.semanticModel = semanticModel;
+            this.currentBlock = currentBlock;
+        }
+
+        public override void VisitBlock(BlockSyntax node)
+        {
+            if (currentBlock == node) // Don't visit nested blocks, that will be done by additional LoggingCallCollector instances.
+            {
+                currentBlock = node;
+                base.VisitBlock(node);
+            }
+        }
+
+        public override void VisitInvocationExpression(InvocationExpressionSyntax node)
+        {
+            if (IsLoggerMethod(node.GetName())
+                && semanticModel.GetSymbolInfo(node).Symbol is IMethodSymbol methodSymbol
+                && methodSymbol.ContainingType.DerivesFromAny(LoggerTypes)
```

Comment:
Shouldn't this be `DerivesOrImplementsAny`?
Since some of the loggers are interfaces.